src/main/java/org/omnifaces/cdi/jsparam/JSParamExtension.java

```diff
@@ -0,0 +1,191 @@
+ * exist, one is created together with a {@link HtmlCommandScript}. The command script is also connected to a action
+ * listener that actually handles the injection logic - being called as soon as the page is loaded.
+ * Once done, the bean instance holding the injection points is checked for all<code>&#64;</code>{@link JSParam}
+ * properties defined in the class. Each property receives the JavaScript and hidden input fields needed to send the
+ * result of the evaluated expressions back to the CDI managed bean.
+ * 
+ * @since 3.3
+ * @author Adam Waldenberg, Ejwa Software/Hosting
+ */
+public class JSParamExtension implements Extension {
+	private static final int GENERATED_ID_NUM_CHARACTERS = 8;
+	private static final String JSPARAM_FORM_NAME = JSParam.class.getName().toLowerCase().replace('.', '-');
+	private static final String ASSIGNMENT_TEMPLATE = "document.getElementById('%s:%s').value = %s;";
+
+	private HtmlBody getDocumentBody() {
+		for (UIComponent c : Faces.getViewRoot().getChildren()) {
+			if (c instanceof HtmlBody) {
+				return (HtmlBody) c;
+			}
+		}
+
+		throw new IllegalStateException(String.format("The docoument %s does not contain a body",
+		                                              Faces.getContext().getViewRoot()));
+	}
+
+	private <T> void initializePageForm(T instance, Field field) {
+		if (Components.findComponent(JSPARAM_FORM_NAME) == null) {
+			final HtmlForm form = new HtmlForm();
+			final HtmlCommandScript commandScript = new HtmlCommandScript();
+
+			commandScript.addActionListener(new JSParamActionListener());
+			commandScript.setName(RandomStringUtils.randomAlphabetic(GENERATED_ID_NUM_CHARACTERS));
+			commandScript.setAutorun(true);
+			form.setStyle("display: none;");
+			form.setId(JSPARAM_FORM_NAME);
+			form.getChildren().add(commandScript);
+			getDocumentBody().getChildren().add(form);
+		}
+	}
+
+	private <T> void addHiddenComponent(T instance, AnnotatedField field) {
+		final HtmlForm form = (HtmlForm) Components.findComponent(JSPARAM_FORM_NAME);
+		final HtmlInputHidden input = new HtmlInputHidden();
+		input.setId(field.getJavaMember().getName());
+		input.getAttributes().put("field", field);
+		form.getAttributes().put("instance", instance);
+		form.getChildren().add(input);
+	}
+
+	private <T> void addScript(T instance, Field field) {
+		final Class<?> beanClass = field.getDeclaringClass();
+		final List<Field> fields = FieldUtils.getFieldsListWithAnnotation(beanClass, JSParam.class);
+		final Field lastField = IteratorUtils.get(fields.iterator(), fields.size() - 1);
+
+		if (field.equals(lastField)) {
+			final List<String> scripts = new ArrayList<>();
+
+			fields.forEach(f -> {
+				final String javascript = f.getAnnotation(JSParam.class).value();
+
+				scripts.add(String.format(ASSIGNMENT_TEMPLATE,
+					JSPARAM_FORM_NAME, f.getName(), javascript)
+				);
+			});
+
+			Components.addScriptToBody(StringUtils.join(scripts, "\n"));
```

Comment:
IMHO JS in the body or even inline JS should be avoided - because Content-Security-Policy. A script should be placed as an external resource in the pages header only.

Reply:
This is true. However, we should probably clean it up in the rest of OmniFaces also then - because it's being used in the code base in the exact same manner. What do you think @BalusC ? This is actually a good point.